packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -167,6 +211,33 @@ export default createRule<Options, MessageId>({
         if (invalidAncestor.type === AST_NODE_TYPES.ReturnStatement) {
           // handle return statement
 
+          if (options.ignoreVoidInVoid) {
+            const functionNode = findFunction(invalidAncestor);
+
+            if (
+              targetNodeFunctionAncestorNodeHasVoidReturnType(invalidAncestor)
+            ) {
+              return;
+            }
+
+            if (functionNode != null) {
+              if (functionNode.type === AST_NODE_TYPES.FunctionDeclaration) {
+                if (functionNode.returnType) {
+                  return;
+                }
+              } else {
+                if (
+                  isValidFunctionExpressionReturnType(functionNode, {
+                    allowTypedFunctionExpressions: true,
+                  }) ||
+                  ancestorHasReturnType(functionNode)
+                ) {
+                  return;
+                }
+              }
+            }
```

Comment:
`targetNodeFunctionAncestorNodeHasVoidReturnType` basically does the same that `findFunction` does + check if the found function has `returnType`. Couldn't we just add `|| functionNode.returnType` to the last if-statement? I think the rule logic won't be changed

```suggestion
            if (functionNode != null) {
              if (functionNode.type === AST_NODE_TYPES.FunctionDeclaration) {
                if (functionNode.returnType) {
                  return;
                }
              } else if (
                  isValidFunctionExpressionReturnType(functionNode, {
                    allowTypedFunctionExpressions: true,
                  }) ||
                  ancestorHasReturnType(functionNode)
                  || functionNode.returnType
                ) {
                  return;
                }
            }
```

Reply:
To remove the function, check the returnType of all functionNodes and execute the `isValidFunctionExpressionReturnType` function only for `functionExpression` and `arrowFn` to avoid errors, so I slightly modified the suggested code to reflect it.